sympy/functions/combinatorial/numbers.py

```diff
@@ -950,9 +957,16 @@ def _eval_rewrite_as_tractable(self, n, m=1, limitvar=None, **kwargs):
         return self.rewrite(polygamma).rewrite("tractable", deep=True)
 
     def _eval_evalf(self, prec):
-        from sympy.functions.special.gamma_functions import polygamma
-        if all(i.is_number for i in self.args):
-            return self.rewrite(polygamma)._eval_evalf(prec)
+        if not all(x.is_number for x in self.args):
+            return
+        n = self.args[0]._to_mpmath(prec)
+        m = (self.args[1] if len(self.args) > 1 else S.One)._to_mpmath(prec)
+        with workprec(prec):
+            if m == 1:
+                res = mp.harmonic(n)
+            else:
+                res = mp.zeta(m) - mp.zeta(m, n+1)
```

Comment:
Why not use recursive evalf on sympy.zeta?

Reply:
Let's say `harmonic().evalf()` looks like this:
```python
    def _eval_evalf(self, prec):
        from sympy.functions.special.zeta_functions import zeta
        if all(i.is_number for i in self.args):
            return self.rewrite(zeta)._eval_evalf(prec)
```

Then one of the test cases I added will fail:
https://github.com/sympy/sympy/blob/96a51c83529162fec3171b451a15764bafba80b3/sympy/functions/combinatorial/tests/test_comb_numbers.py#L296

The rewritten Piecewise expression will not resolve to the special case despite `m` being exactly one and `zeta()` will be numerically evaluated at first argument 1, resulting in a pole error.

```
Traceback (most recent call last):
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/functions/combinatorial/tests/test_comb_numbers.py", line 296, in test_harmonic_evalf
    assert str(harmonic(7.0, 1.0).evalf(n=10)) == '2.592857143'
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/function.py", line 475, in __new__
    result = result.evalf(prec_to_dps(pr))
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/evalf.py", line 1659, in evalf
    result = evalf(self, prec + 4, options)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/evalf.py", line 1501, in evalf
    xe = x._eval_evalf(prec)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/functions/combinatorial/numbers.py", line 962, in _eval_evalf
    return self.rewrite(zeta)._eval_evalf(prec)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/basic.py", line 1943, in rewrite
    return self._rewrite(pattern, rule, method, **hints)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/basic.py", line 1955, in _rewrite
    rewritten = meth(*args, **hints)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/functions/combinatorial/numbers.py", line 917, in _eval_rewrite_as_zeta
    (zeta(m) - zeta(m, n+1), True))
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/function.py", line 475, in __new__
    result = result.evalf(prec_to_dps(pr))
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/evalf.py", line 1659, in evalf
    result = evalf(self, prec + 4, options)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/evalf.py", line 1501, in evalf
    xe = x._eval_evalf(prec)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/function.py", line 593, in _eval_evalf
    v = func(*args)
  File "/usr/local/lib/python3.10/dist-packages/mpmath/functions/zeta.py", line 533, in zeta
    return ctx._zeta(s, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/mpmath/ctx_mp_python.py", line 998, in f
    return ctx.make_mpf(mpf_f(x._mpf_, prec, rounding))
  File "/usr/local/lib/python3.10/dist-packages/mpmath/libmp/gammazeta.py", line 984, in mpf_zeta
    return mpf_zeta_int(to_int(s), prec, rnd)
  File "/usr/local/lib/python3.10/dist-packages/mpmath/libmp/gammazeta.py", line 917, in mpf_zeta_int
    raise ValueError("zeta(1) pole")
ValueError: zeta(1) pole
```

Reply:
@asmeurer is there a way to use the recursive evalf _and_ have Piecewise recognise the special case $m=1.0$ properly?